File: js/30-39/problem31.js
const util = require('../util');

function coinSums() {
  let arr = [];
  for (let twoHundred = 0; twoHundred <= 1; twoHundred++) {
    for (let oneHundred = 0; oneHundred <= 2; oneHundred++) {
      for (let fifty = 0; fifty <= 4; fifty++) {
        for (let twenty = 0; twenty <= 10; twenty++) {
          for (let ten = 0; ten <= 20; ten++) {
            for (let five = 0; five <= 40; five++) {
              for (let two = 0; two <= 100; two++) {
                for (let one = 0; one <= 200; one++) {
                  if ((twoHundred * 200) + (oneHundred * 100) + (fifty * 50) + (twenty * 20) + (ten * 10) + (five * 5) + (two * 2) + (one * 1) > 200) {
                    break;
                  }
                  if ((twoHundred * 200) + (oneHundred * 100) + (fifty * 50) + (twenty * 20) + (ten * 10) + (five * 5) + (two * 2) + (one * 1) === 200) {
                    arr.push([twoHundred + ' x 200', oneHundred + ' x 100', fifty + ' x 50', twenty + ' x 20', ten + ' x 10', five + ' x 5', two + ' x 2', one + ' x 1']);
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  return arr.length;
};

util.timeFunc(coinSums);

File: python/30-39/problem31.py
def coin_sums(n):
    coins = [1,2,5,10,20,50,100,200]

    one = [x * coins[0] for x in range(n // 1 + 1)]
    two = [x * coins[1] for x in range(n // 2 + 1)]
    five = [x * coins[2] for x in range(n // 5 + 1)]
    ten = [x * coins[3] for x in range(n // 10 + 1)]
    twenty = [x * coins[4] for x in range(n // 20 + 1)]
    fifty = [x * coins[5] for x in range(n // 50 + 1)]
    onehundred = [x * coins[6] for x in range(n // 100 + 1)]
    twohundred = [x * coins[7] for x in range(n // 200 + 1)]

    answer = 0

    for a in twohundred:
        for b in onehundred:
            for c in fifty:
                for d in twenty:
                    for e in ten:
                        for f in five:
                            for g in two:
                                for h in one:
                                    new_val = a + b + c + d + e + f + g + h
                                    if new_val > n:
                                        break
                                    if new_val == n:
                                        answer += 1
                                        break
    return answer


if __name__ == '__main__':
    import time
    start_time = time.time()
    print(coin_sums(200))
    print('Time: ' + str(time.time() - start_time) + ' seconds')
